### PR TITLE
exec() kill (dispose/^C) method

### PR DESCRIPTION
**Pull request by Conner1115**
Added a `dispose` method returned by the `exec()` function.  The output (`result`) is now returned as an unresolved promise which can be awaited.

## Usage
```ts
import { experimental, messages } from "@replit/extensions";

const { result: out, dispose } = await experimental.exec({
      args: "tsc --noEmit --watch",
      onOutput: (output) => {
        messages.showConfirm(output);
      },
    });

    setTimeout(() => {
      dispose();
      messages.showError("Stopped");
    }, 3000);

    messages.showWarning((await out).output);
```

### CODE REVIEW

**Comment by lafkpages** [comment]
Is it SIGTERM, SIGKILL or SIGINT?

**Comment by lafkpages** [comment]
So it closes the Goval channel, and Goval handles killing the process(es)?

**Comment by Conner1115** [comment]
No idea tbh.  I think this is crosis rather than goval.

**Comment by lafkpages** [comment]
Well, Crosis is the client and Goval is the server.

File: modules/react-example/App.tsx [hunk 1]
```diff
@@ -12,9 +12,19 @@ export default function App() {
   }
 
   const execute = async () => {
-    const out = await experimental.editor.getPreferences();
+    const { result: out, kill } = await experimental.exec({
```
Comment (lunaroyster) [inline modules/react-example/App.tsx:15]:
```suggestion
    const { result: out, dispose } = await experimental.exec({
```

File: modules/react-example/App.tsx [hunk 2]
```diff
@@ -12,9 +12,19 @@ export default function App() {
...
+      args: "tsc --noEmit --watch",
+      onOutput: (output) => {
+        messages.showConfirm(output);
+      },
+    });
 
-    console.log(out);
+    setTimeout(() => {
+      kill();
+      messages.showError("Killed");
```
Comment (lunaroyster) [inline modules/react-example/App.tsx:24]:
```suggestion
      dispose();
      messages.showError("Killed");
```

File: modules/extensions/src/api/experimental/exec.ts [hunk 3]
```diff
@@ -52,18 +53,26 @@ export async function exec(
     })
   );
 
-  await promise;
+  const result: Promise<SeparatedOutputExecResult | CombinedOutputExecResult> =
+    new Promise((resolve) => {
+      promise.then(() => {
```
Comment (lunaroyster) [inline modules/extensions/src/api/experimental/exec.ts:58]: we want to use the resolved value of this promise to get the exitError (which i also renamed to exitCode in #72)

Comment (Conner1115) [reply modules/extensions/src/api/experimental/exec.ts:58]: Problem with waiting for the promise to be resolved is that you can't dispose the process.  If you somehow get into an infinitely-looping process like `tsc --watch`, the promise will never resolve.  Won't the exit code get returned with the promise?

Comment (lunaroyster) [reply modules/extensions/src/api/experimental/exec.ts:58]: a running process has no exit code, so it makes sense that we never resolve the `result`

also we aren't supplying the exitCode anywhere yet, it's just set to an empty string

Comment (lunaroyster) [inline modules/extensions/src/api/experimental/exec.ts:58]: result

Comment (lunaroyster) [inline modules/extensions/src/api/experimental/exec.ts:58]:
```suggestion
      promise.then((result) => {
```

File: modules/extensions/src/api/experimental/exec.ts [hunk 4]
```diff
@@ -52,18 +53,26 @@ export async function exec(
...
+        if (options.separateStdErr) {
+          resolve({
+            output: outputStr,
+            error: errorStr,
+            exitError: exitCode || null,
```
Comment (lunaroyster) [inline modules/extensions/src/api/experimental/exec.ts:63]:
```suggestion
            exitError: result.exitCode || null,
```

File: modules/extensions/src/api/experimental/exec.ts [hunk 5]
```diff
@@ -52,18 +53,26 @@ export async function exec(
...
+          });
+        } else {
+          resolve({
+            output: outputStr,
+            exitError: exitCode || null,
```
Comment (lunaroyster) [inline modules/extensions/src/api/experimental/exec.ts:68]:
```suggestion
            exitError: result.exitCode || null,
```

**Review by lunaroyster — APPROVED** [review]

**Pull request merged**